### PR TITLE
Fix #319; Adjust initialization of the first and last network layers

### LINKED ISSUE #319: Adjust initialization of the first and last network layers
As the first and last network layers have fewer inputs and outputs than the generic hidden layer, they should, according to Xaiver initialization

correct

draw from a different distributions. Currently they don't, but they should. The goal of this issue is to fix that.

### PR DESCRIPTION
**Pull request by andi4191**
@kdavis-mozilla, @reuben @tilmankamp  : Hi Kelly, Request you to have a look at this and review for comments, if any.

### CODE REVIEW

File: DeepSpeech.py [hunk 1]
```diff
@@ -358,8 +358,8 @@ def BiRNN(batch_x, seq_length, dropout):
     # clipped RELU activation and dropout.
 
     # 1st layer
-    b1 = variable_on_worker_level('b1', [n_hidden_1], tf.random_normal_initializer(stddev=FLAGS.b1_stddev))
-    h1 = variable_on_worker_level('h1', [n_input + 2*n_input*n_context, n_hidden_1], tf.random_normal_initializer(stddev=FLAGS.h1_stddev))
+    b1 = variable_on_worker_level('b1', [n_hidden_1], tf.contrib.layers.xavier_initializer(uniform=False))
```
Comment (tilmankamp) [inline DeepSpeech.py]: Is Xavier also required for the biases?

File: DeepSpeech.py [hunk 2]
```diff
@@ -358,8 +358,8 @@ def BiRNN(batch_x, seq_length, dropout):
...
+    h1 = variable_on_worker_level('h1', [n_input + 2*n_input*n_context, n_hidden_1], tf.contrib.layers.xavier_initializer(uniform=False))
```
Comment (tilmankamp) [inline DeepSpeech.py]: Issue #319 mentioned only first and last layers, but why not using this adaptive approach also for the other layers?
Something to check: Somewhere I read about ReLUs requiring a higher input variance than usual non-linearities, as half of their activation is 0.

File: data/ldc93s1/LDC93S1.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+0 46797 She had your dark suit in greasy wash water all year.
```
Comment (tilmankamp) [inline data/ldc93s1/LDC93S1.txt]: Why is this now part of the repo?

File: DeepSpeech.py (hunk 1 above)
Comment (andi4191) [inline DeepSpeech.py]: After some digging, it seems its not a good idea to initialize bias with xavier_initializer Referenced: https://github.com/tensorflow/tensorflow/issues/3825

Biases can also be initialized as zeros. So, it made sense to me using the same for bias as well. Will revert the change for bias

File: DeepSpeech.py (hunk 2 above)
Comment (andi4191) [inline DeepSpeech.py]: Xavier_initialization assumes linearity in its derivation. So, that may be one of the reason.
First and last layers have fewer inputs.

Are you referring to following paper:
https://arxiv.org/pdf/1502.01852.pdf

File: data/ldc93s1/LDC93S1.txt (hunk 3 above)
Comment (andi4191) [inline data/ldc93s1/LDC93S1.txt]: I will remove this.

File: DeepSpeech.py (hunk 2 above)
Comment (tilmankamp) [inline DeepSpeech.py]: Yes, that was the paper...

Comment (andi4191) [inline DeepSpeech.py]: They are suggesting to initialize their own version of Xavier as sqrt(2/n) where n is the number of inputs.

**Comment by kdavis-mozilla** [comment]
@andi4191 should @tilmankamp take another look at this now?

**Comment by andi4191** [comment]
@kdavis: yeah that will be great.
@tilmankamp: Can you check this out ? I have incorporated the review comments.

**Comment by tilmankamp** [comment]
@andi4191 Have you done some test runs? Any improvements in regards to WER/loss progress?

**Comment by andi4191** [comment]
@tilmankamp : Yeah I ran over ldc93s1 dataset with 50 epochs to test it.

Results:
With above changes:
 FINISHED Optimization - training time: 0:01:06
I Test of Epoch 50 - WER: 0.090909, loss: 8.09268379211, mean edit distance: 0.019231

2nd run
I FINISHED Optimization - training time: 0:01:02
I Test of Epoch 50 - WER: 0.000000, loss: 2.75919222832, mean edit distance: 0.000000

Without changes:
I FINISHED Optimization - training time: 0:01:02
I Test of Epoch 50 - WER: 0.181818, loss: 8.02116012573, mean edit distance: 0.038462

PS: Running over librivox is taking too much time.

**Comment by kdavis-mozilla** [comment]
@andi4191 Merge at will!

**Comment by kdavis-mozilla** [comment]
Just had a thought, which may surprise some, I wonder if this changes the graph and thus disallows us from using older models/checkpoints. 

@andi4191 could you check if checkpoints made before this commit are compatible with those after the commit? I'd just check with something simple like a ldc93s1 model.

**Comment by andi4191** [comment]
@kdavis-mozilla: Changing the initialization shouldn't change the graph, right? (Please correct me if I am incorrect)

I ran the following test:
1. Ran the code over ldc93s1 without fix for #319. The checkpoints were saved.
2. Ran with fix code over the same without removing the previous checkpoint files.
3. Training seems to be fine. Following output were observed.
_D Queue runners started.
D step: 50
D epoch: 50
D target epoch: 50
D steps per epoch: 1
D batches per job: 1
D batches per step: 1
D number of jobs in train set: 1
D number of jobs already trained in first epoch: 0
D Computing Job (ID: 2, worker: 0, epoch: 50, set_name: test)...
D Starting batch...
D Finished batch step 50.
D Sending Job (ID: 2, worker: 0, epoch: 50, set_name: test)...
I Test of Epoch 50 - WER: 0.090909, loss: 9.93421840668, mean edit distance: 0.019231_


Checked the timestamps also of the checkpoint files only the epoch50 file's timestamp was updated (not the epoch 0 file).

Without fix code run was at 12:36. 
Next run with fix was started at 12:38

-rw-r--r--  1 muser  staff       261 Jun 20 12:39 checkpoint
-rw-r--r--  1 muser  staff   2694927 Jun 20 12:37 events.out.tfevents.1497987386.muser-35342
-rw-r--r--  1 muser  staff   2718743 Jun 20 12:39 events.out.tfevents.1497987568.muser-35342
-rw-r--r--  1 muser  staff   1515418 Jun 20 12:39 graph.pbtxt
-rw-r--r--  1 muser  staff         4 Jun 20 12:36 model.ckpt-0.data-00000-of-00002
-rw-r--r--  1 muser  staff  88102292 Jun 20 12:36 model.ckpt-0.data-00001-of-00002
-rw-r--r--  1 muser  staff      1608 Jun 20 12:36 model.ckpt-0.index
-rw-r--r--  1 muser  staff    715681 Jun 20 12:36 model.ckpt-0.meta
-rw-r--r--  1 muser  staff         4 Jun 20 12:39 model.ckpt-50.data-00000-of-00002
-rw-r--r--  1 muser  staff  88102292 Jun 20 12:39 model.ckpt-50.data-00001-of-00002
-rw-r--r--  1 muser  staff      1608 Jun 20 12:39 model.ckpt-50.index
-rw-r--r--  1 muser  staff    715856 Jun 20 12:39 model.ckpt-50.meta

**Comment by kdavis-mozilla** [comment]
@andi4191 Awesome work. Thanks!
